No-frills support for Entity version bumping

DHager opened the pull request:
A cut-down version of DDC-3640 / PR #1378

This PR focuses on the "support code", and does not contain any annotations or other mechanisms for knowing _when_ an entity's version should be bumped in the absence of "direct" modifications. Instead, it mimics other insert/update features with the methods:

```
$em->lock($entity,LockMode::OPTIMISTIC_FORCE_INCREMENT); // Preferred usage

$unitOfWork->lock($entity,LockMode::OPTIMISTIC_FORCE_INCREMENT); 
$unitOfWork->scheduleForVersionBump($entity,$enable); // For internal usage
$unitOfWork->isScheduledForVersionBump($entity);      // For internal usage
$unitOfWork->getScheduledEntityVersionBumps();        // For internal usage
```

These simple calls will allow users to solve tricky use-cases right now with their own custom event-listener code, and serve as a foundation for when DDC-3781 eventually defines an officially-sanctioned method.

Depends on doctrine/dbal#905

Commit:
A cut-down version of DDC-3640 / PR #1378

Commit:
Reordering some code to make the diff from master look cleaner

DHager commented:
Unit tests are failing due to a separate bug in the test-framework, waiting for #1380 to be merged.

Commit:
Document that sometimes preUpdate events can be provided with a blank changeset.

Commit:
Provide a way for outside code to ask UnitOfWork about pending version-bumps.

Document it--alongside similar methods--in the section for onFlush events.

Commit:
Add comment in tests explaining why we need to access UnitOfWork

Refactor UOW-getting so it's a big less cumbersome.

DHager committed:
Merge branch 'master' into version_increment_support

DHager commented:
I'd appreciate any feedback on this before I build too much. Would it help if I put in a test-implementation of an event-listener which implements some child->parent version update logic?

DHager commented:
@zeroedin-bill : Any comments on this as a starting point for [DDC-3781](http://www.doctrine-project.org/jira/browse/DDC-3781)?

DHager commented:
@beberlei , @Ocramius , any opinion on this? I'm willing to sink in quite a bit of time into making  feature for locking on an "aggregate root" entity, but I want to make sure that there's some buy-in on the basic approach and use-case.

DHager commented:
@Ocramius Ping. Is there some other contributor I should be addressing things to?

zeroedin-bill commented:
@DHager Again, entity versioning for optimistic locking is supposed to be read-only to the user. Adding methods for forcibly bumping the versions of arbitrary entities doesn't really meet that test. This is a no-go for inclusion as public API, I think. 

A good start to [DDC-3781](http://www.doctrine-project.org/jira/browse/DDC-3781) would be a set of unit tests that cover the various scenarios where a cascading bidirectional relation between two versioned entities should set the version number.

Ocramius commented:
> These simple calls will allow users to solve tricky use-cases

I don't think this is a good idea in first place: I'd rather design around those issues instead of pushing them down to the tool.

 > Again, entity versioning for optimistic locking is supposed to be read-only to the user

Agree with this.

DHager commented:
> Again, entity versioning for optimistic locking is supposed to be read-only to the user. [...]  This is a no-go for inclusion as public API, I think.

@zeroedin-bill I agree in principle, but consider the context: A quarter of all `public` API methods on `UnitOfWork` are *already* commented as "INTERNAL", and a good portion of the rest are things that "the user" shouldn't be touching either. 

What I'm trying to do here is to follow the existing design-patterns, so that if a *future* "Make UOW API bulletproof" rewrite happens, the same architectural fix (whatever it may be) applies to version-bumping just like it'll have to apply to updates, inserts, deletes, extra-updates, and so on.

> > These simple calls will allow users to solve tricky use-cases right now with their own custom event-listener code, and serve as a foundation for when DDC-3781 eventually defines an officially-sanctioned method.

> I'd rather design around those issues instead of pushing them down to the tool.

@Ocramius I'm don't think I'm parsing you correctly here: It sounds like you're saying that (A) the ORM and its tools should avoid/sidestep multi-entity versioning issues but also (B) user-level code should not be involved either.

DHager commented:
If the event-system is reserved for user-level additions, I'll look into making something a bit like `CommitOrderCalculator`, except focused on discovering sets of "mutually versioned" entities rather than a topological sort of entity-classes.

zeroedin-bill commented:
@DHager If I were you I would consider implementing this as a behavior. Prior to this existing in doctrine2 core, it was implemented externally.

DHager commented:
I don't understand: I thought "behaviors" were a feature of Doctrine 1.x that wasn't carried over.

zeroedin-bill commented:
@DHager Yes, I am referring to those projects as examples. If you need an entity version that you can forcibly bump, you are free to write the mappings, listener, etc to support that behavior, as is done in Atlantic18/DoctrineExtensions for other behaviors.

DHager commented:
> If you need an entity version that you can forcibly bump, you are free to write the mappings, listener, etc to support that behavior

@zeroedin-bill Yes, and this PR is a **prerequisite** for all of that.

zeroedin-bill commented:
@DHager would the semantics in section 3.4.4.1 of the [JPA 2.0 spec](http://download.oracle.com/otn-pub/jcp/persistence-2.0-fr-oth-JSpec/persistence-2_0-final-spec.pdf) work for your purposes?

DHager commented:
@zeroedin-bill :   It seems to me like Doctrine's current behavior could be seen as: "Always make an `OPTIMISTIC` lock, and upgrade that lock to `OPTIMISTIC_FORCE_INCREMENT` ... if and only if the entity has direct changes."

Most of the code in the PR is about bypassing that "if and only if" portion in a "safe" way that doesn't interfere with change-set calculation logic.

So to answer your question, yes, I think `$uow->lock($entity,LockMode::OPTIMISTIC_FORCE_INCREMENT)` would do just as well from a class-API standpoint.

zeroedin-bill commented:
@DHager if an entity has changed state and is to be persisted, the version will be incremented in the process. There is no force increment mechanic in Doctrine's behavior currently. Calling lock($entity, ...OPTIMISTIC_FORCE_INCREMENT) would force entities with no changes to increment their version at the end of the transaction.

Also, I'm definitely not trying to be obstructionist here. Limiting the scope creep of the UoW is important, as it's unmanageable as it is anyway. Even if the UoW were easier to manage, it would still not be good public API I think. Hibernate for example explicitly forbids modifying an entity's version property also. The only method they have for forcibly incrementing it is the lock mode as described above. I think that's probably the best model to follow moving forward.

DHager committed:
Change semantics to use the lock() method and LockMode::OPTIMISTIC_FORCE_UPDATE

DHager commented:
I'm checking in a set of changes for review, but they won't actually run successfully until doctrine/dbal#905 , which adds the `LockMode::OPTIMISTIC_FORCE_INCREMENT` constant.

DHager committed:
Oops, used wrong constant-name in my vendor/ folder, fixed to match other PR.

DHager committed:
Update documentation note about preUpdate vs. versioning

Should make it easier for readers to understand what triggers the exceptional case

DHager committed:
Update unit-test to exercise both LockMode::OPTIMISTIC and LockMode::OPTIMISTIC_FORCE_INCREMENT

(Tests that rely on the difference between the two modes are found elsewhere.)

DHager commented:
@zeroedin-bill The em/uow `lock()` methods should work now, but I still don't see a good mechanism to seal away `UnitOfWork::scheduleForVersionBump()` so that a determined user can't use it. 

I believe it's in the same situation as other `public` methods such as `UnitOfWork::scheduleExtraUpdate()` or `UnitOfWork::scheduleOrphanRemoval()`, so instead I've just given it the same kind of "INTERNAL" commenting to discourage use.

zeroedin-bill commented:
@DHager - Have not had time to review this, will hopefully get to it next week. Thanks.

One note - I'm not sure version bumps should trigger lifecycle callbacks. @Ocramius any thoughts?

Ocramius commented:
@zeroedin-bill no, IMO they shouldn't. It's 99.99% abuse-case anyway.

Marco Pivetta

http://twitter.com/Ocramius

http://ocramius.github.com/

On 18 September 2015 at 19:21, Bill Schaller <notifications@github.com>
wrote:

> @DHager <https://github.com/DHager> - Have not had time to review this,
> will hopefully get to it next week. Thanks.
>
> One note - I'm not sure version bumps should trigger lifecycle callbacks.
> @Ocramius <https://github.com/Ocramius> any thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/doctrine/doctrine2/pull/1456#issuecomment-141512493>.
>

DHager commented:
@Ocramius Wouldn't it be a valid case when using a behavior like Timestampable? It'd just be a little odd for the version to increment but a separate last-modified/last-modified-by field still shows an outdated value.

zeroedin-bill commented:
@DHager The purpose of the version field is *only* *for* *locking*. Timestamps are domain data. The version is entirely for the persistence layer to worry about.

DHager commented:
@zeroedin-bill Do you mean that [Timestampable](https://github.com/Atlantic18/DoctrineExtensions/blob/master/doc/timestampable.md), [Blameable](https://github.com/KnpLabs/DoctrineBehaviors#blameable) etc. are *existing* "abuse-cases" that pervert Doctrine's event system?

Assuming they are legitimately used for direct simple entity changes, I can't see what makes them any less-legitimate for reacting to indirect changes on the "unit of consistency" entity. (e.g. `BillingLineItem` versus `BillingOrder`.)

zeroedin-bill commented:
@DHager Updating the parent entity's timestamp (or any other) field would mark that entity as dirty, and would result in a version increment anyway.

> Suppose that I have a BillingOrder (my desired unit-of-consistency) which owns 20 BillingLineItem entities. The business requirement is "show when the order last changed". I'd rather read one timestamp field on BillingOrder... as opposed to retrieving 21 datetimes that must all be compared to discover which one is the biggest and represents the "most recent subpart changed".

This can be implemented without LockMode::OPTIMISTIC_FORCE_INCREMENT by simply adding a @PreUpdate event to the child entity which updates the parent's last modified timestamp.

Making a version bump with no other associated changes is not in my opinion an update to the entity, and shouldn't trigger the event listener for it. Adding another scenario which triggers the lifecycle events adds unnecessary complexity, and will negatively impact performance. Especially considering that the desired behavior can be achieved more clearly using the preUpdate and prePersist events on the child entity.

DHager commented:
> This can be implemented without LockMode::OPTIMISTIC_FORCE_INCREMENT by simply adding a @PreUpdate event to the child entity which updates the parent's last modified timestamp.

I disagree for two reasons:

First, that's relying on *incidental domain behavior* (the business' need for a date+time data) in order to drive what ought to be a purely persistence-layer concern. Namely, the requirement that a `BillingOrder` is the unit-of-consistency. It is equivalent to what I [wrote before](https://github.com/doctrine/doctrine2/pull/1378#issuecomment-91105655) about a "junk column" being used to *artificially* drive the version number of the parent. 

Secondly, it is an *unreliable* mechanism, due to race-conditions, timestamp-granularity, and synchronization differences between different web-hosts that might be talking to the same database.

DHager commented:
@Ocramius About the lifecycle callbacks... Right now Doctrine has features for supporting Entities and Value Objects. My overall goal is to build on that foundation and extent it to supporting [Aggregate Roots](https://lostechies.com/jimmybogard/2008/05/21/entities-value-objects-aggregates-and-roots/) as well.

Even when the entity functioning as the Aggregate-Root doesn't have *direct* changes, it still bears responsibility for "knowing" that the larger multi-entity structure has been changed. (e.g. `BillingOrder` as the coordination-point for all its `BillingLineItem`s.)

Of the existing lifecycle callsbacks, `preUpdate`/`postUpdate` seem the best fit for what's going on, with the bonus that it's already wired up to occur because the relevant row really is getting a change in the database, even if it's only a version-bump.

DHager commented:
@zeroedin-bill , @Ocramius : IMO the main question to answer involves events, and whether to:

1. Trigger `preUpdate` / `postUpdate` when the "root" entity has its version changed, even if no direct changes were made
2. Create new new pre/post events specifically for versioned-entities which will-get/have-gotten their version values updated, either due to direct changes (normal current optimistic-lock behavior) or indirect ones (new force-increment behavior) or both.
3. Have absolutely no changes to events and see if anyone complaints

DHager commented:
Please let me know what I can do to help move this feature forward. 

Of the three alternatives in my previous post, the first is what the PR "naturally" does due to how UOW is structured, and the second is a compromise to avoid rare BC issues. I dislike the third--no events at all--because I'm certain user-level code should be able to react when something changes the Aggregate.

DHager commented:
Ping...

DHager commented:
... Freaking close button is far too easy to hit...

DHager commented:
I'm willing to do whatever research/dev/testing is needed here, but I need some direction.

DHager commented:
@Ocramius , @zeroedin-bill Weekly ping.

zeroedin-bill commented:
@DHager @Ocramius  I think as of now we'll be putting this off until Doctrine 3.

DHager commented:
@zeroedin-bill @Ocramius Perhaps it could be part of a broader discussion on where/how people draw boundaries between subgraphs of entities, when it comes to versioning and cascading.

DHager commented:
@zeroedin-bill : If there's no room in 2.x, what's the timeline on Doctrine 3? I have a project that could really use this support, and without it I'll be forced to create a meaningless "junk" column on a bunch of entities in order to trick Doctrine.

DHager commented:
@zeroedin-bill Unless Doctrine 3.0 is coming real soon, can I persuade you to reconsider?

This is *not* a big architectural change from how 2.5 already works, but it makes a huge difference in whether the database needs ugly workaround-columns and whether the domain-objects can be persistence-agnostic.

DHager commented:
@Ocramius Is now the right time to start talking about handling "aggregate" versioning in 3.0? Should I open a new PR against the `develop` branch?

Ocramius commented:
@DHager I'm really out of the loop about this patch, so I'll need to re-evaluate it when I have time to go through the issue queue again (end of December, likely)

DHager commented:
@Ocramius At this point I expect "do we want to support this and how" is probably more important than "is this a good patch". 

In an nutshell, implementing an [Aggregate Root](http://martinfowler.com/bliki/DDD_Aggregate.html) which stores the version-counter for a *subgraph* of tightly-related entities. A change within any node of the subgraph, or adding/removing nodes from the subgraph, should be reflected in the root-entity's version.

beberlei commented:
I am not sure now why we didn't implement OPTIMISTIC_FORCE_INCREMENT back in the days, I can't remember seeing it before, but from your use cases and other tickets I have seen it makes a lot of sense to have this.

I am closing this PR here, because the implementation would need some serious rethought, but the idea of adding support for this lock mode is now for next 2.x. No guarantee it gets implemented, but its on the mind: https://github.com/doctrine/orm/issues/3620
